Added SRE Levels (#72)
gravitational/careers: levels/sre.tex

\documentclass{article}

\usepackage[table]{xcolor}
\usepackage[paperwidth=800mm, paperheight=575mm]{geometry}

% Don't show page numbers.
\pagenumbering{gobble}

% Column settings for skills table.
\newcolumntype{C}{p{20em}}

\begin{document}

% Title of matrix.
\begin{center}
\begin{huge}
Site Reliability Engineering (SRE)
\end{huge}
\end{center}

\bigbreak

% Alternating row colors.
\rowcolors{2}{gray!80!gray!50}{gray!50!gray!20}

% Padding for each cell.
{\renewcommand{\arraystretch}{2}

% Skills table.
\begin{center}
\begin{tabular}{|C|C|C|C|C|C|C|C|}
\hline
\rowcolor{blue!60!black!10}
    \textbf{Level/Skill}
    &
    \cellcolor{blue!25}
    \textbf{Product Output}
    &

>     \textbf{Communication/Writing}

Reviewer: It would be great to incorporate RCA/Postmortems into the communication/writing section for SREs.

    &
    \textbf{Networking}
    &
    \textbf{Security}
    &
    \textbf{Systems Engineering}
    &
    \textbf{Site Reliability Engineering (SRE)}
    \\
\hline
    % Level
    1
    &

    % Product Output
    \cellcolor{blue!25}
    Creates a design document based on well-defined scoped requirements and
    implements it.
    &

    % Communication
    Reports progress on a regular basis as required by the team's operational

>     requirements. Actively solicits feedback.Participates on interview panels.

Reviewer 2: We don't have interview panels in the systems levels - don't think we need it here either - although I don't feel strongly about it either way.

    &

    % Networking
    Understands common networking issues and troubleshooting techniques.

    \bigbreak

    Understands different networking layers (OSI model).

    \bigbreak

    Understands and can deploy overlay networks with routing and DNS resolution.
    &

    % Security
    Understands operating systems security principles like iptables,
    users/groups, file permissions, and capabilities. Can apply operating system
    hardening guides like CIS Benchmarks.

    \bigbreak

    Understands basic security principles like SSH keys and TLS certificates.
    &

    % Systems Engineering
    &

    % SRE
    Understands and can make changes to existing Makefiles, shell scripts, and
    Dockerfiles.

    \bigbreak

    Understands declarative configuration and can use tooling like Terraform or
    a Kubernetes operator to make changes to existing infrastructure and
    applications.

    \bigbreak

>     May have certification like Certified Kubernetes Administrator (CKA).

Reviewer: Level 1 is likely to be new grads with little experience, I wouldn't expect any certifications from someone so new to the role.

Also, CKA was pretty tough. I wouldn't expect anything like that until L3 or so.

    \\ [13em]
\hline
    % Level
    2
    &

    % Product Output
    \cellcolor{blue!25}
    Can write high quality user and product focused documentation.
    &

    % Communication
    Provides constructive review on peers' code and design. Helps new team
    members during their first weeks.
    &

    % Networking
    Can independently troubleshoot most network issues.

    \bigbreak

>     Can setup and configure production quality network infrastructure like DNS,

Reviewer 3: I would move the section about IPSec and WireGuard up a level. They are more common knowledge for folks at senior level with more experience. But in the Cloud age, especially at this level, many have not ever interacted with the tech as it has been largely abstracted away.

    Load Balancers, and encrypted networks (like IPsec or WireGuard). Can reason
    about their reliability.
    &

    % Security
    Understands and can apply basic cryptographic principles. Can configure SSH
    and TLS for a server (chooses appropriate key sizes, algorithms, and versions),
    pick strong authentication and authorization primitives, and appropriate
    encryption for data in transit and rest.

    \bigbreak

    Can setup production quality encrypted networking (like IPsec or
    WireGuard). Can reason about their reliability.
    &

    % Systems Engineering
    &

    % SRE

>     Excellent systems troubleshooter.

Reviewer 2: E.g. can diagnose and resolve cascading failures.

    \bigbreak

    Familiar with and can configure alerts for system performance (CPU, memory,

>     disk, networking utilization). Uses modern BPF tools for troubleshooting.

Reviewer 3: I'd change it to "some tools", as some of them are indeed fairly basic. A lot of these tools would fly waaay over the head of lvl 2 because to _successfully_ use them requires the knowledge of what's underneath, like kernel or syscalls, etc.

>     disk, networking utilization). Uses modern BPF tools for troubleshooting.

Reviewer 2: Maybe replace CPU, memory, disk, etc with latency, traffic, errors, and saturation?

    \bigbreak

    Is a member of on-call rotation and can resolve issues outlined in
    runbooks.

    \bigbreak

    Demonstrates knowledge of AWS. May have certification like AWS Certified
    SysOps Administrator.
    \\ [13em]
\hline
    % Level
    3
    &

    % Product
    \cellcolor{blue!25}
    Collaborates with the team to scope requirements, based on good
    understanding of existing longer term product vision and estimates of the
    system design of a feature of a product.
    &

    % Communication
    Supports less experienced peers' technical skills, answering questions and

>     being a resource. Documents and improves team practices.

Reviewer 3: I'd move `Documents and improves team practices` to lvl 2. It is a relatively basic skill and does not always require particularly deep knowledge or experience, especially given that fact that more senior folks could provide feedback.

    &

    % Networking
    Can setup and operate a multi-region infrastructure and networking
    environment. Can reason about it's performance, reliability, and failure modes.

    \bigbreak

>     Has in-depth understanding of container networking. Can write own CNI plugin

Reviewer 3: I'd move the custom CNI sentence to lvl 4.

    utilizing IPSec or WireGuard on Kubernetes.

    \bigbreak

    Understands advanced networking concepts such as network namespaces, NAT

>     traversal, and BPG.

Reviewer 3: nit: s/BPG/BGP?

    &

    % Security

>     Can build secure systems that will pass quality security audit that will

Reviewer 3: I'd broaden this to something along the lines of `Design and document the security posture of the entire stack that will pass quality security audit with few to no critical exceptions` and move this to lvl4. What I mean is that it is one thing to apply some IAM, and yet another to oversee the security of the whole stack in it's entirety and design the setup on each level that is tailored to our specific needs... and accepting certain tradeoffs.

    uncover few to no critical system design errors.

    \bigbreak

>     Can apply security principals when building systems. Can utilize access

Reviewer 3: I would move the IAM / RBAC piece, and the CloudWatch alerts, down do lvl2. These are actually fairly common skills.

    control primitives (like IAM and RBAC) to limit access to infrastructure.
    Understands secret life cycle management in production environments. Understand
    API authentication systems, can reason about trade-offs between use of JWT,
    OIDC, and mTLS. Can setup alerting on security systems like CloudWatch.

    \bigbreak

    Understands advanced operating system security concepts. Can utilize
    Mandatory Access Control (MAC) systems like as SELinux or AppArmor. Understands
    container security. Can utilize control groups and namespaces to isolate and
    application.
    &

    % Systems Engineering

>     &

Reviewer 3: I think lvl 3 SREs _must_ be able to code, and we _must_ make it explicit. Does not have to be anything fancy, but lvl 3 is someone who "graduated" from bash scripting. Maybe something along the lines of "Ability to write production-ready code for simple services or automation. Code includes unit and integration testing. Supplies clear contribution instructions."

Side note: no explicit coding skills requirements sort of contradicts a requirement to implementat a custom CNI for k8s, hence my suggestion WRT that.

    % SRE

>     Can use declarative languages like Terraform to build and manage infrastructure.

Reviewer: I would swap this with the "Excellent Systems Troubleshooter" in the previous level.

Learning Terraform is easier than becoming an excellent systems troubleshooter IMO, because the latter only comes with experience.

Reviewer 3: ➕ definitely move Terraform down to lvl 2.

however, I think troubleshooting in itself is a skill, i.e. learning to ask for help at the right moment is. The sooner SREs learn this skill, the better. so I'd leave something at level 2 WRT troubleshooting. Maybe similar to network, we could say "Able to troubleshoot many systems problems independently".

    \bigbreak

> 

Reviewer: I wonder if we should add something about time-series DBs, which is pretty much an industry standard at this point.

Knowing the data model and query language is an important skill in order to be able to effectively ask the system questions about how how it is performing.

    Demonstrates advanced knowledge of Kubernetes deployments (CRDs, ingress
    controllers, Load Balancers).

    \bigbreak

    Can build and maintain reliable production CI/CD pipelines for build, test,
    and release.

    \bigbreak

    Demonstrates advanced knowledge of AWS. May have certification like AWS
    Certified Solutions Architect.
    \\ [13em]
\hline
    % Level
    4
    &

    % Product
    \cellcolor{blue!25}
    Leads the implementation of the isolated feature/improvement that
    measurably and significantly impacts business outcomes from gathering
    requirements to getting to the market stage.
    &

    % Communication
    Writes technical articles/blog posts, delivers tech and lightning talks

>     representing the company's technical vision.

Reviewer 3: I would add something along the lines of "Defines or helps to establish SLI/SLO for services and entire products, guides the data collection and setup of the SLI/SLO monitoring". Words are again failing me, but this is an important part of being SRE, and at it gets exponentially harder with scale (both product and team).

    to provide service level load balancing, multi-region connectivity, and
    observability.
    &

    % Security
    Writes technical articles on security aspects of the system, implements
    significant security product innovations in the area delivered to customers.

    \bigbreak

    Understands and can apply advanced cryptographic principles. Understands
    hashing (including for anonymization), when to use symmetric and asymmetric
    cryptography, cipher modes, and TLS versions.

    \bigbreak

    Understands and can apply advanced network security principles. Understands
    data extrusion prevention. Understands DDoS mitigation. Understands how to
    monitor systems for rootkits and can deploy mitigation strategies when a system
    is under attack.
    &

    % Systems Engineering
    Can write data-race and deadlock free code. Can implement safe and
    concurrent and/or parallel systems using minimum amount of shared state,
    granular locking - systems that are easy to read, extend and troubleshoot.
    &

    % SRE
    Can build and operate large scale, stable, and reliable production platform
    environments like Kubernetes.

    \bigbreak

>     Not expected to find any faults or provide any further suggestions in the
>     design document submitted by an Level 4 engineer.

Reviewer 2: ```suggestion
    Design documents should be free of faults or require any further suggestions.
```

    \\ [13em]
\hline
    % Level
    5
    &

    % Product
    \cellcolor{blue!25}

>     Leads the implementation of a new product line or significant part of the

Reviewer 3: Do we realistically expect SREs to work on product features? As in, I'd love to personally, but I have seen this only once on practice.

    product to deliver it to the market in collaboration with all other teams.
    &

    % Communication
    Writes advanced technical articles/blog posts, gaining significant industry
    traction or delivers technical talks on major conferences representing the
    company's vision.
    &

    % Networking
    &

    % Security
    Researches and designs new security systems and protocols.
    &

    % Systems Engineering
    Can implement production grade systems leveraging advanced low-level and/or
    novel components like eBPF, control groups, or Noise Protocol Framework.
    &

    % SRE

>     Understands and use use advanced system performance troubleshooting
>     techniques like ptrace, strace, flamegraphs, or writing custom bpftrace
>     programs.

Reviewer 2: ```suggestion
    Understands and uses advanced system performance troubleshooting
    techniques like ptrace, strace, flamegraphs, or writing custom bpftrace
    programs.
```

    \bigbreak

>     Can deploy and operate databases at large scale. Understands index

Reviewer 3: I'd move "basic" database stuff, like "can operate databases in production" to lvl 4. And maybe even failover/replication, i.e. "ability to configure and troubleshoot replication for failover and horizontal scaling purposes".

>     Can deploy and operate databases at large scale. Understands index
>     compaction, failover, sharing, and query performance analysis.

Reviewer 2: ```suggestion
    Can deploy and operate databases at large scale. Understands index
    compaction, failover, sharding, and query performance analysis.
```

    \bigbreak

    Can build advanced monitoring and anomaly detection systems.
    \\ [13em]
\hline
    % Level
    6
    &

    % Product
    \cellcolor{blue!25}
    Designs new data structures and algorithms solving relevant business

>     problems and creating competitive advantage for the company.

Reviewer 2: I see this is from the systems data-structures levels. Not sure it belongs here. The "solving relevant business problems" is an important piece though. Perhaps we could spin this to designing HA low-latency geo-distributed systems or equivalent.
